add core solidity deep dive blogpost (#158)
ethereum/solidity-website: src/posts/2025-11-14-core-solidity-deep-dive.md

## A Note on Syntax

Most of the work to date has been focused on the design and implementation of the type system, and the
associated code generation pipeline down to Yul. In order to avoid getting bogged down in
bikeshedding around syntax, and with the desire to validate our core ideas with a working
implementation as soon as possible, we moved ahead with a provisional syntax. You can expect
extensive changes before release. Our current intention is to eventually closely match the syntax of Classic Solidity. For any new syntax, you should expect that the final version will feel
closer to languages like TypeScript or Rust.

## New Language Features

Core Solidity takes ideas from pure functional programming languages (e.g. Haskell, Lean), as well
as modern systems languages (e.g. Rust, Zig). We are extending Solidity with the following new
features:

- Algebraic datatypes (also known as sum / product types) and pattern matching
- Generics / parametric polymorphism

> - Traits / typeclasses

Reviewer: ```suggestion
- Traits / type classes
```

Reviewer: Same for "type system" vs "typesystem".

```

The function `unpack_bools` implements a custom ABI decoding. It is a higher-order function which decorates an input function that takes three individual `bool`s and returns any type by extracting the arguments from the right-most three bits. This is an example which is impossible to implement in Classic Solidity, even with [modifiers](https://docs.soliditylang.org/en/latest/contracts.html#function-modifiers), since they cannot change the arguments passed to the wrapped function.

We also support the definition of (non-recursive) anonymous functions using the `lam` keyword.
Functions defined in this way can capture values available in the defining scope. As an example
consider this testing utility that counts the number of times an arbitrary function is called:

```js
forall T U . function count_calls(fn : (T) -> U) -> (memory(word), (T) -> U) {
    let counter : memory(word) = allocate(32);
    return (counter, lam (a : T) -> {
        counter += 1;
        return fn(a);
    });
}
```

Our implementation here is similar to systems languages like Rust and C++: the compiler produces a
unique type for each anonymous function that contains the capture, and these unique types are

> made callable by making them instances of the `invokable` type class (similar to the [`Fn`](https://doc.rust-lang.org/std/ops/trait.Fn.html)

Reviewer: Don't our type class names start with capitals? I tried to check on solcore, but I'm getting mixed results, not sure which ones are relevant.

```suggestion
made callable by making them instances of the `Invokable` type class (similar to the [`Fn`](https://doc.rust-lang.org/std/ops/trait.Fn.html)
```

trait in Rust). This approach is runtime gas efficient.

### Type inference

Core Solidity supports the inference of types in almost any position. Annotations are usually only
needed when considered desirable for readability or understanding. Inference is decidable,
and the situations in which ambiguities requiring annotation can occur are very limited. This lets
us solve a lot of the syntactic clutter required when writing Classic Solidity. As an example,
consider the following Classic Solidity definition:

For example, assigning an expression to a variable in Classic Solidity can often result in redundant annotation if those
types are already present in the expression being assigned:

```js

> (bytes memory a, bytes memory b) = abi.decode(input, (bytes, bytes))

Reviewer: Missing semicolon.
```suggestion
(bytes memory a, bytes memory b) = abi.decode(input, (bytes, bytes));
```

There's one more below.

implement to guard against accidental leakage of references to compile time memory. We are also
investigating what kind of compile time specific primitives we might want to add, and whether we
want to expand the language capabilities around reflection.

We will publish more on our designs once they stabilise. We want to make sure the needs of the
community are met here: if you have concrete real world use cases in mind for this feature, we would
be very interested to hear them.

## SAIL, Desugaring, and the Standard Library

In addition to expanding the surface language, the transition to Core Solidity will also introduce a
new user accessible mid level IR: SAIL (Solidity Algebraic Intermediate Language). This is the
"Core" in Core Solidity. It's the most minimal language that lets us
express the full range of high-level language constructs found in Classic Solidity. It consists of the following primitive constructs:

- Functions
- Contracts
- Assembly (Yul) blocks
- SAIL variable introduction and assignment
- A short circuiting if-then-else expression
- Algebraic datatypes & pattern matching
- Type classes
- Generics

> A SAIL variable is conceptually the same as a [Yul variable](https://docs.soliditylang.org/en/latest/yul.html#variable-declarations). The compiler will associate an EVM stack slot to it.

Reviewer: In blog posts we should generally link to the current version of the docs. The links can break and content can change over time.
```suggestion
A SAIL variable is conceptually the same as a [Yul variable](https://docs.soliditylang.org/en/v0.8.30/yul.html#variable-declarations). The compiler will associate an EVM stack slot to it.
```

There are a few more links like this in the post.

a community driven RFC-style process for changes to the high-level language and standard library.

### A Userspace abi.encode

Let's look at how SAIL can be used to implement high-level Core Solidity features. `abi.encode` is a complicated
and highly generic function that Classic Solidity provides as a compiler builtin. A full in-language
implementation would not be possible in Classic Solidity due to the recursive nature of the ABI
specification and the resulting infinite number of expressible types. The implementation presented
here is relatively concise, but does make use of some more advanced patterns and features. We
want to emphasise that existing users of Solidity will be able to be productive and make use of
their existing knowledge without having to concern themselves with these kind of low level internal
details. At the same time, we hope that advanced users and library authors will be excited by the new
potentials these features enable.

For presentation purposes we restrict ourselves to the fragment required to encode `uint256`.

#### uint256

To begin we will construct the type `uint256`. In Classic Solidity the definition of this type and

> it's associated operations are all built-in language constructs. In SAIL, it is defined entirely in-language as a simple wrapper around a `word`. We also define a `Typedef` instance for it:

Reviewer: ```suggestion
its associated operations are all built-in language constructs. In SAIL, it is defined entirely in-language as a simple wrapper around a `word`. We also define a `Typedef` instance for it:
```

```js
data bytes;
```

Notice that in this construction of pointers and data locations, the data location is attached to
the type (instead of the variable binding as it is in Classic), allowing for the definition of e.g.
memory structures containing references to storage.

#### The Proxy type

The last piece of machinery required for `abi.encode` is the `Proxy` type:

```js
data Proxy(T) = Proxy;
```

As with the `memory` definition, the type parameter here is phantom, but unlike memory `Proxy`
carries no additional information at runtime. It exists only as a marker type that lets us pass
information around at compile time. Types like this are completely zero cost (i.e. they are
completely erased at runtime and do appear in the final compiled program at all).

Although somewhat esoteric, `Proxy` is very useful and gives us a lot of control over type inference
and instance selection without needing to pass data at runtime where it is not needed. It is often
used in both Haskell (where it is also called `Proxy`) and Rust (`std::marker::PhantomData`).

#### abi.encode

Now we are ready to implement Classic Solidity's `abi.encode` in SAIL. We start by defining a

> typeclass for ABI related metadata, note that since this class does not need to care about the

Reviewer: ```suggestion
Now we are ready to implement Classic Solidity's `abi.encode` in SAIL. We start by defining a
typeclass for ABI related metadata. Note that since this class does not need to care about the
```

actual value of the type being passed to it, we use a `Proxy` to keep our implementation as lean as
possible.

```js
forall T . class T:ABIAttribs {

>     // how many bytes should be used for the head portion of the abi encoding of `T`

Reviewer: ```suggestion
    // how many bytes should be used for the head portion of the ABI encoding of `T`
```

There are also a few instances more of `abi` and `yul` without any capitalization in the post.

    function headSize(ty : Proxy(T)) -> word;
    // whether or not `T` is a fully static type
    function isStatic(ty : Proxy(T)) -> bool;
}

instance uint256:ABIAttribs {
    function headSize(ty : Proxy(uint256)) -> word { return 32; }
    function isStatic(ty : Proxy(uint256)) -> bool { return true; }
}
```

Now we define another class that handles the low level encoding into memory. The class presented
here contains some extraneous details needed for encoding compound and dynamic types that are not be
necessary for the simple `uint256` encoding we are implementing now. We present the full complexity

> to demonstrate that we have the machinery required for this harder cases.

Reviewer: ```suggestion
to demonstrate that we have the machinery required for these harder cases.
```

```js
// types that can be abi encoded
forall T . T:ABIAttribs => class T:ABIEncode {
    // abi encodes an instance of T into a memory region starting at basePtr
    // offset gives the offset in memory from basePtr to the first empty byte of the head

>     // tail gives the index in memory of the first empty byte of the tail

Reviewer: ```suggestion
    // tail gives the position in memory of the first empty byte of the tail
```

Reviewer: Also note how `abi` in `abi encodes an instance ...` looks like a parameter name. This could use backticks for variables, proper capitalization and full stops.

to postfix types). We are also considering reworking or replacing some features that have proven to
be problematic or limiting in practice (e.g. try/catch, libraries, function pointers, data
locations). Users can expect that some moderate changes may be required to adapt their code that
makes use of these features. Code that makes heavy use of inheritance will of course require the
largest changes.

We will be investigating the potential for automated upgrades, and if reliable and robust
implementations are possible expect to ship such tools at release.

Avoiding a Python 2 -> Python 3 style split is top of mind, and we believe that upgrades should be
manageable and that it will be possible to carry them out in an incremental manner.

## The Road to Production

This section outlines our current thinking on achieving production readiness and our strategy for
making such deep changes to the language in a safe way. Please note that this is a tentative plan,
and may be subject to extensive change. We are not yet in a position where we feel confident about
committing to concrete timelines. We will provide more details as
we get closer to a production implementation.

> We have a prototype implemented in a [separate repository](https://github.com/argotorg/solcore). We

Reviewer: ```suggestion
We have a prototype implemented in a separate repository: [solcore](https://github.com/argotorg/solcore). We
```
